Validate checkpoint file format in find_last_step

pyrytakala opened the pull request:
Fixes `find_last_step` to properly validate checkpoint file format using regex, preventing crashes when non-standard files exist in the checkpoint directory.

Example crash:
```
# Given files: model_000100.pt, model_000200.pt, model_000200_backup.pt
# Processing model_000200_backup.pt:
split("_")[-1]  # → "backup.pt"
split(".")[0]   # → "backup"
int("backup")   # → ValueError: invalid literal for int() with base 10: 'backup'
```

pyrytakala committed:
Fix find_last_step crash on checkpoint files with extra underscores

svlandeg reviewed (commented):
Hi, thanks for the PR!

Just out of curiosity, how does a file like `model_000200_backup.pt` end up in your directory? I don't think the current scripts do anything like that, right?

Anyway, if we want to make this more robust, we should simply refactor L118 only, there's no need for 10 additional code lines 🙂 . We could simply rewrite L118 into

```
checkpoint_files = [f for f in os.listdir(checkpoint_dir) if re.search(r'model_(\d+)\.pt$', f)]
```
and then we'd apply the more stringent/correct pattern immediately. 

I don't see a way to do this directly with `glob.glob()` as that doesn't support `*` for a variable sequence like `[0-9]*`. (and repeating `[0-9]` 6 times in the pattern would be pretty ugly)

pyrytakala committed:
Refactor find_last_step to use os.listdir with regex filtering

Replace glob.glob() with os.listdir() + regex filtering as suggested by reviewer.
This filters invalid checkpoint files (like model_000200_backup.pt) at the source
instead of globbing then filtering, making the code simpler and more efficient.

pyrytakala commented:
Thanks for the suggestion @svlandeg . I've implemented your suggestion, much cleaner!

To answer your question: the current scripts don't create backup files, but I wanted to make the function robust against edge cases where someone might manually create or rename checkpoint files

pyrytakala committed:
Update nanochat/checkpoint_manager.py

Co-authored-by: Sofie Van Landeghem <svlandeg@users.noreply.github.com>

svlandeg committed:
do int conversation before taking the max

svlandeg reviewed (approved):
I've now also included @kuizhiqing's suggestion to apply the `max` operator after converting to `int`. This will come down to the exact same thing, because the file names are formatted as `"model_{step:06d}.pt"`, with each `int` encoded using the same number of digits, so the `max` will in fact work correctly on the strings as well. It's just that semantically it does make more sense to me to apply `int` first, and then the `max` - it's also more robust.

So all in all this PR looks good to me - all changes are mostly just for robustness of the code, but they are minimal changes that do no obfuscate any behaviour IMO.
